### PR TITLE
Handle resolving key, issuer references

### PR DESCRIPTION
**Pull request by cipherboy**
The API context will usually have a user-specified reference to the key.
This is either the literal string "default" to select the default key,
an identifier of the key, or a slug name for the key. Here, we wish to
resolve this reference to an actual identifier that can be understood by
storage.

Also adds the missing Name field to keys.

`Signed-off-by: Alexander Scheel <alex.scheel@hashicorp.com>`

---

This depends on four unwritten methods and thus can't be built yet:

```
builtin/logical/pki/storage.go:83:18: undefined: getKeysConfig
builtin/logical/pki/storage.go:91:15: undefined: listKeys
builtin/logical/pki/storage.go:151:18: undefined: getIssuersConfig
builtin/logical/pki/storage.go:159:18: undefined: listIssuers
```

In `fetchCAInfo`, we'd call `resolveIssuerReference` and then call `fetchCertBundleByIssuerId`.

### CODE REVIEW

File: builtin/logical/pki/storage.go [hunk 1]
```diff
@@ -76,6 +77,45 @@ func writeKey(ctx context.Context, s logical.Storage, key key) error {
...
+			return keyId("key-read"), err
+		}
+
+		if key.Name == reference {
+			return key.ID, nil
+		}
+	}
+
+	// Otherwise, we must not have found the key.
+	return keyId("not-found"), fmt.Errorf("unable to find PKI key for reference: %v", reference)
```
Comment (cipherboy) [inline builtin/logical/pki/storage.go:116]: This should probably be a UserError along with the one below?

Comment (stevendpclark) [reply builtin/logical/pki/storage.go:116]: Possibly, again it's a convention thing where all the existing codebase within pki only checked for two error types in error handling code (InternalError and UserError). I *believe* I fixed all those handlers but there were a surprising number of them.

**Comment by cipherboy** [comment]
Manually merged. :-)

**Comment by cipherboy** [comment]
This PR was merged in #15277. See that PR and the relevant docs PR #15238 for more information about this change.

**Pull request closed without merging**